thetis/exner_eq.py

```diff
@@ -111,8 +111,17 @@ def residual(self, solution, solution_old, fields, fields_old, bnd_conditions=No
 
         # bnd_conditions are the shallow water bcs, any boundary for which
         # nothing is specified is assumed closed
+
         for bnd_marker in (bnd_conditions or []):
-            f += -self.test*(fac*qbx*self.n[0] + fac*qby*self.n[1])*self.ds(bnd_marker)
+            no_contr = False
+            keys = [*bnd_conditions[bnd_marker].keys()]
+            values = [*bnd_conditions[bnd_marker].values()]
+            for i in range(len(keys)):
+                if keys[i] != 'elev' and float(values[i]) == 0.0:
+                    no_contr = True
+
+            if not no_contr:
+                f += -self.test*(fac*qbx*self.n[0] + fac*qby*self.n[1])*self.ds(bnd_marker)
```

Comment:
So if I understand this correctly, for closed boundary you want to omit the boundary term on line 124? So currently the code does this (un applies for uv and flux conditions as well):
```
1 unprescribed : open
2 elev : open
3 un != 0 : open
4 un == 0 : closed
5 elev and un != 0 : open
6 elev and un == 0 : closed
```
I'm not sure if this is right. To be consistent with SWE in Thetis, case 1 should be closed. Is it necessary to differentiate the closed condition with un==0? Can you not just do
```
1 unprescribed : closed
2 elev : open
3 un : open
4 elev and un : open
```

Reply:
I'm not sure I entirely understand but this is what I think is happening in these lines of code:
1 unprescribed - closed - no exner bc
2 un = 0 - closed - no exner bc (irrespective of other conditions)
3 un != 0 - open - exner bc
4 elev and no un/un !=0 - open - exner bc
It is necessary to create a catch for cases 1 and 2 because there are advantages to using case 2 in the other equations. You are right that because of this "if clause" from the perspective of the exner equation cases 1 and 2 are the same

Reply:
Hmm, so line 124 means that you evaluate the boundary flux with the vector (qbx, qby). Thus this is an "open" boundary. If the boundary term is omitted, it is identically zero, and therefore we have a "closed" boundary.
So the logic tree is
```
1 un prescribed and un == 0  -> no_contr = True -> omit bnd term -> closed
2 all other cases: no_contr = False -> include bnd term -> open
```
Do you agree with this?

Reply:
Yes I agree with this, apart from case 1 hold for both un and flux (ie. if flux exists and flux==0 then closed)

Reply:
Also here let's redo the boundaries once bendy elements work